src/v/storage/tiered_storage.cc

```diff
@@ -0,0 +1,516 @@
+#include "storage/tiered_storage.h"
+
+#include "bytes/iobuf_istreambuf.h"
+#include "cloud_storage/types.h"
+#include "config/configuration.h"
+#include "hashing/xx.h"
+#include "json/json.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "model/record_batch_types.h"
+#include "model/timestamp.h"
+#include "s3/client.h"
+#include "s3/error.h"
+#include "storage/log_reader.h"
+#include "storage/logger.h"
+#include "storage/ntp_config.h"
+#include "storage/parser.h"
+#include "utils/gate_guard.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/iostream-impl.hh>
+#include <seastar/core/iostream.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/core/temporary_buffer.hh>
+#include <seastar/util/log.hh>
+
+#include <absl/container/btree_map.h>
+#include <boost/algorithm/string/detail/sequence.hpp>
+#include <boost/lexical_cast.hpp>
+
+#include <chrono>
+#include <exception>
+#include <variant>
+
+static ss::logger recovery_logger("recovery");
+
+namespace storage {
+using namespace std::chrono_literals;
+
+static constexpr ss::lowres_clock::duration download_timeout = 300s;
+static constexpr ss::lowres_clock::duration initial_backoff = 200ms;
+
+/// Partition that we're trying to fetch from S3 is missing
+class missing_partition_exception final : public std::exception {
+public:
+    explicit missing_partition_exception(const ntp_config& ntpc)
+      : _msg(ssx::sformat(
+        "missing partition {}, rev {}", ntpc.ntp(), ntpc.get_revision())) {}
+
+    explicit missing_partition_exception(
+      const cloud_storage::remote_manifest_path& path)
+      : _msg(ssx::sformat("missing partition s3://{}", path)) {}
+
+    const char* what() const noexcept override { return _msg.c_str(); }
+
+private:
+    ss::sstring _msg;
+};
+
+partition_recovery_manager::~partition_recovery_manager() {
+    vassert(_gate.is_closed(), "S3 downloader is not stopped properly");
+}
+
+ss::future<> partition_recovery_manager::stop() { co_await _gate.close(); }
```

Comment:
The cloud_storage::remote is shared between recovery and archival. So during shutdown the remote is stopped which also stops he recovery.

I think that I need to add cleanup after such shutdown though. If we will stop recovery this way after restart it could try to start with that partial data and it will use config batch from the downloaded log which will prevent raft group from bootstrapping correctly.